applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js

```diff
@@ -0,0 +1,178 @@
+'use strict';
+'require uci';
+'require view';
+'require poll';
+'require ui';
+'require rpc';
+
+
+return view.extend({
+	callGetOlsrVizData: rpc.declare({
+		object: 'olsrvizinfo',
+		method: 'getolsrvizdata'
+	}),
+
+	fetch_jsoninfo: function () {
+		var jsonreq4 = '';
+		var json;
+		var data;
+		var self = this;
+		return new Promise(function (resolve, reject) {
+			L.resolveDefault(self.callGetOlsrVizData(), {})
+				.then(function (res) {
+					json = res;
+					data = json.jsonreq4;
+					resolve([data]);
+				})
+				.catch(function (err) {
+					console.error(err);
+					reject([null]);
+				});
+		});
+	},
+
+	action_olsr_viz: function () {
+		var self = this;
+		return new Promise(function (resolve, reject) {
+			self
+				.fetch_jsoninfo()
+				.then(function ([data]) {
+					var result = { viz_data: data };
+					resolve(result);
+				})
+				.catch(function (err) {
+					reject(err);
+				});
+		});
+	},
+
+	load: function () {
+		var self = this;
+		poll.add(function () {
+			self.render();
+		}, 15);
+	},
+	render: function () {
+		var viz_res;
+		var self = this;
+		return this.action_olsr_viz()
+			.then(function (result) {
+				viz_res = result.viz_data;
+
+				var cssScript = E('script', {
```

Comment:
Oh dear, that's some artifact :) You can replace the CSS injection with just:

```javascript
document.querySelector('head').appendChild(E('style', { 'type': 'text/css' }, [
  '.label {color:black;background-color:white}'
]));
```

To load the `olsr-viz.js` script I suggest to inject the `<script>` tag from within the `load` function, then wrap the onload event callback into a promise to defer `render()` until the script is loaded. This way you can subsequently skip the `var script = '<script language="JavaScript1.2" type="text/javascript">viz_setup("RSIFrame", "main","nodes", "edges");viz_update();</script>';` part and simply call `viz_setup(...)` and `viz_update()` synchronously.

Within the `load` function:

```javascript
return new Promise(function(resolveFn, rejectFn) {
    var script = E('script', { 'type': 'text/javascript' });
    script.onload = resolveFn;
    script.onerror = rejectFn;
    script.src = L.resource('olsr-viz.js');
    document.querySelector('head').appendChild(script);
});
```

Reply:
Okay I'll fix them all, Thanks for the review!